src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -4784,25 +4779,73 @@ void MacroAssembler::string_equals(Register a1, Register a2,
 
   // Check for short strings, i.e. smaller than wordSize.
   subs(cnt1, cnt1, wordSize);
-  br(Assembler::LT, SHORT);
-  // Main 8 byte comparison loop.
-  bind(NEXT_WORD); {
-    ldr(tmp1, Address(post(a1, wordSize)));
-    ldr(tmp2, Address(post(a2, wordSize)));
+  br(LT, SHORT);
+
+  if (!UseSimpleStringEquals) {
+    Label STUB, B16, B24;
+
     subs(cnt1, cnt1, wordSize);
-    eor(tmp1, tmp1, tmp2);
-    cbnz(tmp1, DONE);
-  } br(GT, NEXT_WORD);
-  // Last longword.  In the case where length == 4 we compare the
-  // same longword twice, but that's still faster than another
-  // conditional branch.
-  // cnt1 could be 0, -1, -2, -3, -4 for chars; -4 only happens when
-  // length == 4.
-  ldr(tmp1, Address(a1, cnt1));
-  ldr(tmp2, Address(a2, cnt1));
-  eor(tmp2, tmp1, tmp2);
-  cbnz(tmp2, DONE);
-  b(SAME);
+    br(LE, B16);
+    subs(cnt1, cnt1, wordSize);
+    br(LE, B24);
+    subs(cnt1, cnt1, wordSize);
+    br(GT, STUB);
+
+    ldr(rscratch1, Address(post(a1, wordSize)));
+    ldr(rscratch2, Address(post(a2, wordSize)));
+    eor(rscratch1, rscratch1, rscratch2);
+    cbnz(rscratch1, DONE);
+
+    bind(B24);
+    ldr(rscratch1, Address(post(a1, wordSize)));
+    ldr(rscratch2, Address(post(a2, wordSize)));
+    eor(rscratch2, rscratch1, rscratch2);
+    cbnz(rscratch2, DONE);
+
+    bind(B16);
+    ldr(rscratch1, Address(post(a1, wordSize)));
+    ldr(rscratch2, Address(post(a2, wordSize)));
+    eor(rscratch1, rscratch1, rscratch2);
+    cbnz(rscratch1, DONE);
+
+    ldr(rscratch1, Address(a1, cnt1));
+    ldr(rscratch2, Address(a2, cnt1));
+    eor(rscratch2, rscratch1, rscratch2);
+    cbnz(rscratch2, DONE);
+    b(SAME);
+
```

Comment:
No, we are not going to do all this unrolling at the call site.

Reply:
* Why we unrolling the loop ?
*  It is found that these codes degrades performance 
 ```
    br(LE, B16);
    subs(cnt1, cnt1, wordSize);
    br(LE, B24);
    subs(cnt1, cnt1, wordSize);
```
 We unrolls the loop to remove these comparsion.

Reply:
* Why we unroll the loop ?
*  It is found that these codes degrades performance 
 ```
    br(LE, B16);
    subs(cnt1, cnt1, wordSize);
    br(LE, B24);
    subs(cnt1, cnt1, wordSize);
```
 We unroll the loop to remove these comparsion.

Reply:
* Why we unroll the loop ?
*  It is found that these codes degrades performance 
 ```
    br(LE, B16);
    subs(cnt1, cnt1, wordSize);
    br(LE, B24);
    subs(cnt1, cnt1, wordSize);
```
 We unroll the loop to remove these comparsion.

Reply:
But we need to balance performance against code size, given that this is expanded frequently.

Reply:
Nick is right. This unrolling at the call site is not going to be accepted.